tap-agent/src/tap/receipt_storage_adapter.rs

```diff
@@ -0,0 +1,598 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    num::TryFromIntError,
+    ops::{Bound, RangeBounds},
+};
+
+use alloy_primitives::Address;
+use async_trait::async_trait;
+
+use sqlx::{postgres::types::PgRange, types::BigDecimal, PgPool};
+use tap_core::adapters::receipt_storage_adapter::ReceiptStorageAdapter as ReceiptStorageAdapterTrait;
+use tap_core::{
+    tap_manager::SignedReceipt,
+    tap_receipt::{ReceiptCheck, ReceivedReceipt},
+};
+use thiserror::Error;
+use tracing::error;
+
+#[derive(Debug)]
+pub struct ReceiptStorageAdapter {
+    pgpool: PgPool,
+    allocation_id: Address,
+    sender: Address,
+    required_checks: Vec<ReceiptCheck>,
+}
+
+#[derive(Debug, Error)]
+pub enum AdapterError {
+    #[error("Error in ReceiptStorageAdapter: {error}")]
+    AdapterError { error: String },
+}
+
+impl From<TryFromIntError> for AdapterError {
+    fn from(error: TryFromIntError) -> Self {
+        AdapterError::AdapterError {
+            error: error.to_string(),
+        }
+    }
+}
+
+impl From<sqlx::Error> for AdapterError {
+    fn from(error: sqlx::Error) -> Self {
+        AdapterError::AdapterError {
+            error: error.to_string(),
+        }
+    }
+}
+
+impl From<serde_json::Error> for AdapterError {
+    fn from(error: serde_json::Error) -> Self {
+        AdapterError::AdapterError {
+            error: error.to_string(),
+        }
+    }
+}
+
+/// convert Bound`<u64>` to Bound`<BigDecimal>`
+fn u64_bound_to_bigdecimal_bound(bound: Bound<&u64>) -> Bound<BigDecimal> {
+    match bound {
+        Bound::Included(val) => Bound::Included(BigDecimal::from(*val)),
+        Bound::Excluded(val) => Bound::Excluded(BigDecimal::from(*val)),
+        Bound::Unbounded => Bound::Unbounded,
+    }
+}
+
+/// convert RangeBounds`<u64>` to PgRange`<BigDecimal>`
+fn rangebounds_to_pgrange<R: RangeBounds<u64>>(range: R) -> PgRange<BigDecimal> {
+    // Test for empty ranges. Because the PG range type does not behave the same as
+    // Rust's range type when start > end.
+    if match (range.start_bound(), range.end_bound()) {
+        (Bound::Included(start), Bound::Included(end)) => start > end,
+        (Bound::Included(start), Bound::Excluded(end)) => start >= end,
+        (Bound::Excluded(start), Bound::Included(end)) => start >= end,
+        (Bound::Excluded(start), Bound::Excluded(end)) => start >= end || *start == end - 1,
+        _ => false,
+    } {
+        // Return an empty PG range.
+        return PgRange::<BigDecimal>::from(BigDecimal::from(0)..BigDecimal::from(0));
+    }
+    PgRange::<BigDecimal>::from((
+        u64_bound_to_bigdecimal_bound(range.start_bound()),
+        u64_bound_to_bigdecimal_bound(range.end_bound()),
+    ))
+}
+
+#[async_trait]
+impl ReceiptStorageAdapterTrait for ReceiptStorageAdapter {
+    type AdapterError = AdapterError;
+
+    /// We don't need this method in TAP Agent.
+    async fn store_receipt(&self, _receipt: ReceivedReceipt) -> Result<u64, Self::AdapterError> {
```

Comment:
What component calls this function and `update_receipt_by_id`?

Reply:
It the TapManager from tap_core: https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/3d247c06380038d0a871a7427ce00e0406909160/tap_core/src/tap_manager/manager.rs#L80-L111.

But we don't need those since the storage of receipts is done only in `service`, and in a mush simplified way, so it doesn't even leverage the TapManager at all.